// Example/models/UserEmail.js
module.exports= UserEmail

UserEmailMapper= require('Example/mappers/UserEmail')



function UserEmail(data) {

    this.id= data.id || 0
    this.value= data.value || null

    if (data.updatedAt) {
        this.updatedAt= Date(data.updatedAt)
    }

    if (data.deleted) {
        this.deleted= !!data.deleted
    }
    if (data.deletedAt) {
        this.deletedAt= Date(data.deletedAt)
    }

}



UserEmail.query= function (user) { return function (done) {

    var err, emails= []

    setTimeout(function () {
        emails= UserEmailMapper.query(user)
        done(err, emails)
    }, 37)

}}



UserEmail.push= function (user, data) { return function (done) {

    var err, emails= []

    data= data || []
    data.map(function (email) {
        emails.push(
            new UserEmail(email)
        )
    })

    setTimeout(function () {
        UserEmailMapper.push(user, emails)
        done(err, emails)
    }, 73)

}}



UserEmail.patch= function (user, data) { return function (done) {

    var err, emails= []

    data= data || []
    data.map(function (email) {
        emails.push(
            new UserEmail(email)
        )
    })

    setTimeout(function () {
        UserEmailMapper.patch(user, emails)
        done(err, emails)
    }, 73)

}}



UserEmail.del= function (user) { return function (done) {

    var err

    setTimeout(function () {
        UserEmailMapper.del(user)
        done(err)
    }, 73)

}}
