// Example/models/UserPhone.js
module.exports= UserPhone

UserPhoneMapper= require('Example/mappers/UserPhone')



function UserPhone(data) {

    this.id= data.id || 0
    this.value= data.value || null

    if (data.updatedAt) {
        this.updatedAt= Date(data.updatedAt)
    }

    if (data.deleted) {
        this.deleted= !!data.deleted
    }
    if (data.deletedAt) {
        this.deletedAt= Date(data.deletedAt)
    }

}



UserPhone.ExistsError= function () { }



UserPhone.query= function (user) { return function (done) {

    var err, phones= []

    setTimeout(function () {
        phones= UserPhoneMapper.query(user)
        done(err, phones)
    }, 37)

}}



UserPhone.push= function (user, data) { return function (done) {

    var err, phones= []

    data= data || []
    data.map(function (phone) {
        phones.push(
            new UserPhone(phone)
        )
    })

    setTimeout(function () {
        UserPhoneMapper.push(user, phones)
        done(err, phones)
    }, 73)

}}



UserPhone.check= function (user) { return function (done) {

    var err, phones= user.phones || []

    setTimeout(function () {
        UserPhoneMapper.check(user, phones)
        done(err, phones)
    }, 73)

}}



UserPhone.checkCode= function (code) { return function (done) {

    var err

    setTimeout(function () {
        code= UserPhoneMapper.checkCode(code)
        done(err, code)
    }, 73)

}}



UserPhone.patch= function (user, data) { return function (done) {

    var err, phones= []

    data= data || []
    data.map(function (phone) {
        phones.push(
            new UserPhone(phone)
        )
    })

    setTimeout(function () {
        UserPhoneMapper.patch(user, phones)
        done(err, phones)
    }, 73)

}}



UserPhone.del= function (user) { return function (done) {

    var err

    setTimeout(function () {
        UserPhoneMapper.del(user)
        done(err)
    }, 73)

}}
