// Example/models/User.js
module.exports= User

UserMapper= require('Example/mappers/User')



function User(data) {

    this.id= data.id || 0
    this.name= data.name

    if (data.updatedAt) {
        this.updatedAt= Date(data.updatedAt)
    }

    if (data.deleted) {
        this.deleted= !!data.deleted
    }
    if (data.deletedAt) {
        this.deletedAt= Date(data.deletedAt)
    }

}



User.ExistsError= function () { }



User.query= function (query) { return function (done) {

    var err, users= []

    setTimeout(function () {
        users= UserMapper.query()
        done(err, users)
    }, 37)

}}



User.get= function (data) { return function (done) {

    var err, user= new User(data)

    setTimeout(function () {
        user= UserMapper.get(user)
        done(err, user)
    }, 37)

}}



User.post= function (data) { return function (done) {

    var err, user= new User(data)

    setTimeout(function () {
        UserMapper.post(user)
        done(err, user)
    }, 73)

}}



User.patch= function (data) { return function (done) {

    var err, user= new User(data)

    setTimeout(function () {
        UserMapper.patch(user)
        done(err, user)
    }, 73)

}}



User.del= function (data) { return function (done) {

    var err, user= new User(data)

    setTimeout(function () {
        UserMapper.del(user)
        done(err, user)
    }, 37)

}}
