// Example/mappers/UserPhone.js
module.exports= {

    id: 0,
    models: {},

    query: function (parent) {
        var models= []
        Object.keys(this.models[parent.id] || []).map(function (key) {
            models.push(this.models[parent.id][key])
        }, this)
        return models
    },

    push: function (parent, models) {
        models.map(function (model) {
            model.id= ++this.id
            model.updatedAt= Date()
            this.models[parent.id]= this.models[parent.id] || {}
            this.models[parent.id][model.id]= model
        }, this)
    },

    patch: function (parent, models) {
        var forInsert= []
        var forUpdate= []
        var forDelete= []
        models.map(function (model) {
            if (model.id) {
                if (model.deleted) {
                    forDelete.push(model)
                } else {
                    forUpdate.push(model)
                }
            } else {
                forInsert.push(model)
            }
        }, this)
        forDelete.map(function (model) {
            if (this.models[parent.id][model.id]) {
                delete this.models[parent.id][model.id]
                model.updatedAt= model.deletedAt= Date()
            }
        }, this)
        forUpdate.map(function (model) {
            if (this.models[parent.id][model.id]) {
                this.models[parent.id][model.id]= model
                model.updatedAt= Date()
            }
        }, this)
        forInsert.map(function (model) {
            model.id= ++this.id
            model.updatedAt= Date()
            this.models[parent.id][model.id]= model
        }, this)
    },

    del: function (parent) {
        delete this.models[parent.id]
    }

}
