// Example/mappers/UserEmail.js
module.exports= {

    id: 0,
    models: {},

    codes: {},
    codesIdx: {},

    query: function (parent) {
        var models= []
        Object.keys(this.models[parent.id] || []).map(function (key) {
            models.push(this.models[parent.id][key])
        }, this)
        return models
    },

    push: function (parent, models) {
        models.map(function (model) {
            model.id= ++this.id
            model.updatedAt= Date()
            this.models[parent.id]= this.models[parent.id] || {}
            this.models[parent.id][model.id]= model
        }, this)
    },

    check: function (parent, models) {
        models.map(function (model) {
            if (this.models[parent.id][model.id].verifiedAt) return
            model.verifiedAt= this.models[parent.id][model.id].verifiedAt= null
            model.updatedAt= this.models[parent.id][model.id].updatedAt= Date()
            this.codes[parent.id]= this.codes[parent.id] || {}
            var code= this.codes[parent.id][model.id]= Math.random()
            this.codesIdx[code]= this.models[parent.id][model.id]
            console.log('EMAIL CODE:', code)
        }, this)
    },

    checkCode: function (code) {
        this.codesIdx[code].verifiedAt= Date()
        delete this.codesIdx[code]
        return code
    },

    patch: function (parent, models) {
        var forInsert= []
        var forUpdate= []
        var forDelete= []
        models.map(function (model) {
            if (model.id) {
                if (model.deleted) {
                    forDelete.push(model)
                } else {
                    forUpdate.push(model)
                }
            } else {
                forInsert.push(model)
            }
        }, this)
        forDelete.map(function (model) {
            if (this.models[parent.id][model.id]) {
                delete this.models[parent.id][model.id]
                model.updatedAt= model.deletedAt= Date()
            }
        }, this)
        forUpdate.map(function (model) {
            if (this.models[parent.id][model.id]) {
                this.models[parent.id][model.id]= model
                model.updatedAt= Date()
                model.verifiedAt= null
            }
        }, this)
        forInsert.map(function (model) {
            model.id= ++this.id
            this.models[parent.id][model.id]= model
            model.updatedAt= Date()
            model.verifiedAt= null
        }, this)
    },

    del: function (parent) {
        delete this.models[parent.id]
    }
}
